### PR TITLE
feat(acl): added subtype to access collections

### PR DESCRIPTION
**Pull request by jdalsem**
This helps identifying what a access collection is used for.

BREAKING CHANGE:
The group metadata has been removed in favor of a access collection
subtype.

- [x] docs
- [x] migrating docs
- [x] manual testing
- [x] tests for api

### CODE REVIEW

File: engine/classes/Elgg/Database/AccessCollections.php [hunk 1]
```diff
@@ -530,14 +530,22 @@ public function canEdit($collection_id, $user_guid = null) {
...
+	public function create($name, $owner_guid = 0, $subtype = null) {
 		$name = trim($name);
 		if (empty($name)) {
 			return false;
 		}
+		
+		if (isset($subtype)) {
+			$subtype = trim($subtype);
+			if (strlen($subtype) > 255) {
+				return false;
```
Comment (jeabakker) [inline engine/classes/Elgg/Database/AccessCollections.php]: maybe an log an error why this failed

**Comment by jdalsem** [comment]
this PR is preparation for moving friends into an ACL... need to finish this first and then can move on to creating FriendsACLs

File: engine/schema/migrations/20171106100916_add_acl_subtype.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+		
+		// remove the migrated group_acl metadata
+		$this->query("
+			DELETE FROM {$prefix}metadata md
+			INNER JOIN {$prefix}entities e ON md.entity_guid = e.guid
+			INNER JOIN {$prefix}{$table->getName()} acl ON md.value = acl.id
+			WHERE md.name = 'group_acl'
+			AND e.type = 'group'
+		");
+    }
```
Comment (jeabakker) [inline engine/schema/migrations/20171106100916_add_acl_subtype.php]: indentation if this file seems of

File: mod/groups/tests/write_access.php [hunk 3]
```diff
@@ -1,78 +0,0 @@
...
-		elgg_set_ignore_access($ia);
-
-		$this->group->leave($this->user);
-
-		$original_page_owner_guid = (elgg_instanceof($original_page_owner)) ? $original_page_owner->guid : 0;
-		elgg_set_page_owner_guid($original_page_owner_guid);
-
-	}
-
-}
```
Comment (jeabakker) [inline mod/groups/tests/write_access.php]: please also remove registration of this file for unit test in `start.php`

**Review by jeabakker — DISMISSED** [review]

File: mod/groups/tests/write_access.php (hunk 3 above)
Comment (jdalsem) [reply mod/groups/tests/write_access.php]: done

File: mod/friends_collections/views/default/collections/listing/owner.php [hunk 4]
```diff
@@ -7,11 +7,11 @@
  */
 
 $entity = elgg_extract('entity', $vars);
-if (!$entity) {
+if (!($entity instanceof \ElggEntity)) {
```
Comment (jeabakker) [inline mod/friends_collections/views/default/collections/listing/owner.php]: why not `ElggUser`?

File: views/default/input/access.php [hunk 5]
```diff
@@ -104,7 +104,10 @@
 }
 
 if ($show_override_notice) {
-	$vars['data-group-acl'] = $container->group_acl;
+	$acls = $container->getOwnedAccessCollections(['subtype' => 'group_acl']);
+	if ($acls) {
+		$vars['data-group-acl'] = $acls[0]->id;
+	}
```
Comment (jeabakker) [inline views/default/input/access.php]: shouldn't this be done in the groups plugin with a `view_vars` hook

**Review by jeabakker — COMMENTED** [review]
I see a lot of duplicate code in the groups plugin, whouldn't it be better to create a helper function to get the group ACL?

File: mod/friends_collections/views/default/collections/listing/owner.php (hunk 4 above)
Comment (jdalsem) [reply mod/friends_collections/views/default/collections/listing/owner.php]: there are no user entity specific functions used in this view or other checks on the type

**Review by jeabakker — APPROVED** [review]

**Comment by jdalsem** [comment]
> I see a lot of duplicate code in the groups plugin, whouldn't it be better to create a helper function to get the group ACL?

this has been taken care of

**Pull request merged**